### ISSUE TITLE
kv: major performance regression in kv workload on single node GCE after txn.go changes

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#30208]
It appears that #28185 caused a big performance regression. Does not show up on macs, but happens on GCE - which might be because the workload's default concurrency is much higher there.

```
bin/workload run kv --read-percent 95 --duration 30s
```

Before:
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   16.0s        0         311615        19499.6      1.8      1.4      4.5      5.8     35.7  read
   16.0s        0          16540         1035.0     12.9     12.6     22.0     27.3     39.8  write
```

After:
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         367987        12265.9      2.8      2.9      5.8      7.1     37.7  read
   30.0s        0          19597          653.2     20.1     19.9     33.6     37.7     50.3  write
```

The perf diff doesn't appear to happen with `--read-percent 100`, which would indicate that we need to focus on `--read-percent 0`, but I haven't done more tests yet.

One first suspicion is that perhaps this workload only does 1PC txns, in which case we're not spinning up heartbeat loops that we weren't before?

cc @jordanlewis

**Comment by jordanlewis** [comment]
Another good thing to keep in mind is that the performance regression mostly goes away if you configure kv to split the keyspace first, with `workload run kv --splits=1000`

**Comment by nvanbenschoten** [comment]
> One first suspicion is that perhaps this workload only does 1PC txns, in which case we're not spinning up heartbeat loops that we weren't before?

Did you mean `s/not/now/`? I thought we avoided heartbeat loops for 1PC txns?

**Comment by jordanlewis** [comment]
The commit description implies that we used to make heartbeat loops for 1PC txns, but no longer:

```
One behavior change introduced by this patch is that heartbeat loops are
no longer started for (what the TCS hopes will be) 1PC txns. The
motivation was concern over the price of spawning a (shortlived)
heartbeat goroutine per txn in the 1PC-heavy "kv" workload.
Another one is that the TxnCoordSender doesn't inherit the old Txn logic
for swallowing errors on rollbacks. Instead, we're relying on a recent
server change to not return errors on rollbacks when the txn record is
missing - which was the reason for said swallowing.
```

Perhaps that analysis is mistaken in this case? 

The query that is run is `UPSERT INTO kv (k, v) VALUES ($1, $2)`. I'm not sure if that's eligible for 1PC,  maybe someone else can chime in?

**Comment by nvanbenschoten** [comment]
As long as there aren't any secondary indices we should absolutely be running that as a 1PC txn. That's worth confirming though.

**Comment by andreimatei** [comment]
Indeed, we do *not* start heartbeats for 1PC txns. I guess I'm glad I put that optimization in at the time, although now I don't know what to suspect for this regression. I also don't know what to make of it going away with the presplits. Much intrigued.

**Comment by andreimatei** [comment]
I ran some experiments. Results are very weird. It looks like the only case where I can see a big regression is the reads-95%, no splits case. Everything else (all writes, all reads, 95% reads with splits) the numbers are close.

```
with my PR: 69fa9f7cff371ce00594770cebf772886ff9d445
without my PR: ff70e8df8ba715d28972ae7f17aa1698aa489117

bin/workload run kv --read-percent 95 --duration 30s
with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         489161        16305.0      2.2      1.8      4.7      6.0     33.6  read
   30.0s        0          25370          845.6     14.8     14.2     26.2     29.4     39.8  write

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         606104        20202.8      1.7      1.3      4.2      5.5     35.7  read
   30.0s        0          31853         1061.7     12.7     12.1     22.0     27.3     39.8  write


bin/workload run kv --read-percent 0 --duration 30s

with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         143323         4777.4     10.1      8.9     21.0     27.3     37.7  write

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  30.0s        0         144819         4827.2      9.9      8.9     19.9     27.3     37.7  write


bin/workload run kv --read-percent 100 --duration 30s

with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
   30.0s        0        1463343        48777.1      1.0      0.9      2.1      4.5     35.7

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
   30.0s        0        1524871        50828.0      0.9      0.8      2.0      4.5     32.5

bin/workload run kv --read-percent 50 --duration 30s

with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         119091         3969.6      0.8      0.7      1.7      2.6     22.0  read
   30.0s        0         118777         3959.1     11.3     10.5     19.9     25.2     32.5  write

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         119874         3995.7      0.8      0.7      1.7      2.6     24.1  read
   30.0s        0         119978         3999.2     11.2     10.5     19.9     25.2     30.4  write


bin/workload run kv --read-percent 95 --duration 30s --splits=1000

with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  30.0s        0        1236113        41202.2      0.8      0.7      1.6      4.5     28.3  read
  30.0s        0          65341         2178.0      6.8      6.3     12.1     15.7     35.7  write

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0        1283666        42787.5      0.8      0.7      1.5      4.5     37.7  read
   30.0s        0          66927         2230.8      6.8      6.3     12.1     15.7     44.0  write


bin/workload run kv --read-percent 0 --duration 30s

with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  30.0s        0         133069         4435.5     10.8     10.0     21.0     26.2     32.5  write
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         140838         4694.5     10.2      9.4     19.9     27.3     33.6  write

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   30.0s        0         142480         4749.2     10.1      9.4     19.9     27.3     33.6  write
```

**Comment by andreimatei** [comment]
One observation I have is cpu usage difference when running kv-95. Without the PR in question, `top` reports crdb using around 900% CPU, and with the PR it goes down to ~630% CPU (on my 24-core GCE worker). So it would appear that we scale less than we used to.
Looking at the cpu profiles and blocking profiles I don't see anything obvious yet. I'm attaching before and after profiles here too if someone wants to take a poke. (Btw, the "raw" blocking profile endpoint seems to generate them in a text format instead of the usual (proto?) format; how do I tell `go tool pprof` to open one of these?)
[profiles.zip](https://github.com/cockroachdb/cockroach/files/2381016/profiles.zip)

**Comment by andreimatei** [comment]
Today on the same machine I can't reproduce more than a 15% difference on kv95. I think I need to test this on an exclusive GCE machine.

**Comment by nvanbenschoten** [comment]
I think I found a lead here. Switching back and forth between CPU profiles with and without this change revealed that [`connExecutor.handleAutoCommit`](https://github.com/cockroachdb/cockroach/blob/69fa9f7cff371ce00594770cebf772886ff9d445/pkg/sql/conn_executor_exec.go#L1264) jumped from taking **.28%** of the profile to **4.81%**. I still don't know what to make of the regression only happening on single-range benchmarks, but this is the kind of fallout I would expect from a change like #28185.

<img width="1187" alt="screen shot 2018-09-15 at 3 29 23 pm" src="https://user-images.githubusercontent.com/5438456/45589925-2e213e00-b8fc-11e8-818a-37a074b17a83.png">

**Comment by andreimatei** [comment]
Might be a very good find, Nathan! Although by itself it doesn't make any of the numbers make more sense. But who knows what 2nd order effect this behavior makes, so I'll focus on this for a bit.
Inspecting the traces, it would appear that the `handleAutoCommit` in question is dealing with read-only transactions. Before my change, there was an optimization that didn't make it forward.
It read:
```
// finishReadonlyLocked provides a fast-path for finishing a read-only
// transaction without going through the overhead of creating an
// EndTransactionRequest only to not send it.
```

So, we were completely eliding all the work done for these commits, whereas now we're not. We're still eliding actually sending the `EndTransaction`, but that happens at a lower level now.
When I've done the change in question, I considered all the work in between to be negligible, but maybe I was wrong.

One immediate improvement is taking care of the `fmt.Errorf()` call. That's just a stupid mistake - we're attempting to increment a duration metric for these read-only transactions that don't have a start time.

**Comment by jordanlewis** [comment]
I think I filed an issue for that last point a long time ago - #28424

**Comment by andreimatei** [comment]
Good news - if I hackily elide all the work for committing r/o txns, all the performance difference goes away. Why it does, I still don't know. The work in question, when not elided, accounts for ~5% of a CPU profile's samples and ~6% of memory allocations (obj count). How can that translate into 50% throughput difference? I see a small sliver of `runtime.gcAssistAlloc`; I don't know if I should make anything of it, but perhaps it's a tell tale of some sort.
The other question is why the perf diff is not there when doing 100% reads. I think maybe the answer here is that the benchmark is fooling itself when it doesn't perform any writes. I suspect the code just ends up attempting to always read a sentinel key '0' which doesn't exist (even if the db has otherwise been populated), and this makes the results pretty somewhat meaningless.

Now focusing on the work that's being elided, it largely falls into three categories wrt to both CPU usage and allocations - TCS context annotation, creating a batch response and cloning the txn for it, and metrics updating.

**Comment by andreimatei** [comment]
I've opened #30448, which takes away much of the performance regression.

I think I've fooled myself above when I said that if I hackily elide all the work for committing r/o txns, all the performance difference goes away. It is not all, but it is a majority.

**Comment by andreimatei** [comment]
Here's the results of #30448, as applied to `69fa9f7cff371ce00594770cebf772886ff9d445` (so, as they relate to this regression).

"with" my PR: 69fa9f7cff371ce00594770cebf772886ff9d445
"without" my PR: ff70e8df8ba715d28972ae7f17aa1698aa48911

Tested on two machines: a 24 cores gce worker and a 96 cores one. On the 24 core machine (the only one I could really get a regression cliff to show up, and even there quite inconsistently), the reads QPS goes up from 13kqps to 20kqps (it was 21.7kqps before)

```
 bin/workload init kv; bin/workload run kv --read-percent 95 --duration 100s --concurrency=48

24 cores
=============

with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  100.0s        0        1289039        12890.2      2.7      2.5      5.8      6.8     41.9  read
  100.0s        0          67709          677.1     20.2     19.9     33.6     39.8     48.2  write


without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  100.0s        0        2176947        21769.3      1.5      0.9      3.9      5.2     37.7  read
  100.0s        0         115185         1151.8     12.8     12.1     22.0     28.3     56.6  write


elide-pool_commits-optimized_metrics:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  100.0s        0        1987268        19872.5      1.7      1.3      4.2      5.5     35.7  read
  100.0s        0         104618         1046.2     13.4     13.1     23.1     28.3     50.3  write



96 cores
=============
with:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  100.0s        0        1691904        16918.9      2.0      1.6      4.7      6.0     58.7  read
  100.0s        0          89175          891.7     15.1     14.7     27.3     31.5     58.7  write

without:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  100.0s        0        2085526        20855.1      1.6      1.0      4.2      5.5     56.6  read
  100.0s        0         109857         1098.6     13.0     12.6     23.1     27.3     52.4  write

elide-pool_commits-optimized_metrics:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  100.0s        0        1938448        19384.3      1.8      1.3      4.5      5.8     67.1  read
  100.0s        0         102184         1021.8     13.5     13.1     23.1     28.3     71.3  write
```

**Comment by awoods187** [comment]
I'm seeing increased performance on TPC-C now!

**Comment by andreimatei** [comment]
Calling this fixed by #30448
